WEB-INTEGRATION.md

```diff
@@ -0,0 +1,742 @@
+  method [\[GEOLOCATION\]](https://w3c.github.io/geolocation/)
+- A number of async methods in
+  [\[ENTRIES-API\]](https://wicg.github.io/entries-api/)
+
+Some of these APIs started out as legacy APIs that took completion callbacks,
+and then they were changed to return a promise – e.g. `BaseAudioContext`’s
+`decodeAudioData()` method. For those APIs, the callback’s context would behave
+similarly to other async completion callbacks, and the promise rejection context
+would behave similarly to other promise-returning web APIs (see below).
+Similarly, the WebIDL-based callback wrapping is sufficient, and there are no
+meaningful alternatives to consider.
+
+### Callbacks run as part of an async algorithm
+
+These APIs always invoke the callback to run user code as part of an
+asynchronous operation that they start, and which affects the behavior of the
+operation. If the callbacks are always run asynchronously, then they should also
+use the context in which the API was called:
+
+- [`Document`](https://dom.spec.whatwg.org/#document):
+  [`startViewTransition()`](https://drafts.csswg.org/css-view-transitions-2/#dom-document-startviewtransition)
+  method [\[CSS-VIEW-TRANSITIONS-1\]](https://drafts.csswg.org/css-view-transitions-1/)
+- [`LockManager`](https://w3c.github.io/web-locks/#lockmanager):
+  [`request()`](https://w3c.github.io/web-locks/#dom-lockmanager-request) method
+  [\[WEB-LOCKS\]](https://w3c.github.io/web-locks/)
+
+You can see these APIs as implementable in JS like this, and they should behave the same:
+```js
+async function api(callback) {
+  await doSomething();
+  await callback();
+  await doSomethingElse();
+}
+```
+
+Some APIs take callbacks that could be seen as running as part of an async
+algorithm, but the callbacks might also run in response to some other web API,
+which can have a different context. In those cases, the callbacks should still
+run in the context in which the web API that started the algorithm was called.
+However, it might also be useful in some cases to make the synchronous context
+available to the callback, for example by passing an `AsyncContext.Snapshot` as
+an additional argument. Some examples:
+
+- The underlying [source](https://streams.spec.whatwg.org/#underlying-source-api)
+  / [sink](https://streams.spec.whatwg.org/#underlying-sink-api) /
+  [transform](https://streams.spec.whatwg.org/#transformer-api) APIs for streams
+  [\[STREAMS\]](https://streams.spec.whatwg.org). Other than the `start` method,
+  which is always invoked synchronously when the stream is constructed, the
+  other underlying API methods are invoked in response to operations on the
+  stream. These operations might have a synchronous cause such as JS code
+  calling `reader.read()`, or an asynchronous cause such as JS code calling
+  `await new Response(stream).text()`, and it might be useful to provide the
+  context for such causes to the callback – for example, by adding an extra
```

Comment:
When might it be useful in this way? I'm trying to anticipate how likely this future extension would be.